sources/slack/__init__.py

```diff
@@ -79,78 +104,56 @@ def users_resource() -> Iterable[TDataItem]:
+        write_disposition="append",
+    )
+    def messages_resource(
         created_at: dlt.sources.incremental[DateTime] = dlt.sources.incremental(
             "ts",
             initial_value=start_dt,
             end_value=end_dt,
             allow_external_schedulers=True,
         ),
     ) -> Iterable[TDataItem]:
-        """Yield all messages for a given channel as a DLT resource.
+        """
+        Yield all messages for a set of selected channels as a DLT resource. Keep blocks column without normalization.
 
         Args:
-            channel_data (Dict[str, Any]): The channel data.
             created_at (dlt.sources.incremental[DateTime]): The incremental created_at field.
 
         Yields:
             Iterable[TDataItem]: A list of messages.
         """
         start_date_ts = ensure_dt_type(created_at.last_value, to_ts=True)
         end_date_ts = ensure_dt_type(created_at.end_value, to_ts=True)
-        params = {
-            "channel": channel_data["id"],
-            "oldest": start_date_ts,
-            "latest": end_date_ts,
-        }
-
-        for page_data in api.get_pages(
-            resource="conversations.history",
-            response_path="$.messages[*]",
-            params=params,
-            datetime_fields=MSG_DATETIME_FIELDS,
-            context={"channel": channel_data["id"]},
-        ):
-            yield page_data
-
-    if selected_channels:
-        fetch_channels = [
-            c
-            for c in channels
-            if c["name"] in selected_channels or c["id"] in selected_channels
-        ]
-    else:
-        fetch_channels = channels
-
-    def table_name_func(channel_name: str, payload: TDataItem) -> str:
-        """Return the table name for a given channel and payload."""
-        table_type = payload.get("subtype", payload["type"])
-        return f"{channel_name}_{table_type}"
-
-    # for each channel, create a messages resource
-    for channel in fetch_channels:
-        channel_name = channel["name"]
-        table_name = partial(table_name_func, channel_name)
-        yield dlt.resource(
-            get_messages_resource,
-            name=channel_name,
-            table_name=table_name,
-            primary_key=("channel", "ts"),
-            write_disposition="append",
-        )(channel)
-
-    # It will not work in the pipeline or tests because it is a paid feature, it raises a
+        for channel_data in fetched_selected_channels:
+            params = {
+                "channel": channel_data["id"],
+                "oldest": start_date_ts,
+                "latest": end_date_ts,
+            }
+
+            for page_data in api.get_pages(
+                resource="conversations.history",
+                response_path="$.messages[*]",
+                params=params,
+                datetime_fields=MSG_DATETIME_FIELDS,
+                context={"channel": channel_data["id"]},
+            ):
+                yield from page_data
```

Comment:
yield whole pages, it's more efficient. it's better to have a for-loop in the "replies_resource" transformer

Reply:
https://dlthub.com/docs/reference/performance#yield-pages-instead-of-rows

Reply:
O, thanks, I was planning to check if this approach is less efficient, I'll rewrite this. Thanks for the link!